File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Projection/EnrollmentList/EnrollmentListProjectionHandler.cs
using CreditCardEnrollment.Common.Projection;
using CreditCardEnrollment.CreditCard.Enrollment.Aggregate;
using CreditCardEnrollment.CreditCard.Enrollment.Event;
using CreditCardEnrollment.CreditCard.Product.Event;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;

public class EnrollmentListProjectionHandler : ProjectionHandler {
    private readonly EnrollmentRepository _enrollmentRepository;
    private readonly ProductNameRepository _productNameRepository;

    public EnrollmentListProjectionHandler(
        EnrollmentRepository enrollmentRepository,
        ProductNameRepository productNameRepository) {
        _enrollmentRepository = enrollmentRepository;
        _productNameRepository = productNameRepository;
    }

    public override void Project(Common.Event.Event @event) {
        switch (@event) {
            case ProductDefined productDefined:
                _productNameRepository.Save(new ProductName {
                    Id = productDefined.AggregateId,
                    Name = productDefined.Name
                });
                break;

            case EnrollmentRequested enrollmentRequested:
                _enrollmentRepository.Save(new Enrollment {
                    Id = enrollmentRequested.AggregateId,
                    UserId = enrollmentRequested.UserId,
                    ProductId = enrollmentRequested.ProductId,
                    RequestedDate = enrollmentRequested.RecordedOn,
                    Status = EnrollmentStatus.Requested.ToString()
                });
                break;

            case EnrollmentAccepted enrollmentAccepted:
                var acceptedEnrollment = _enrollmentRepository.FindOneById(enrollmentAccepted.AggregateId) ?? throw new InvalidOperationException();
                acceptedEnrollment.Status = EnrollmentStatus.Accepted.ToString();
                acceptedEnrollment.ReviewedOn = enrollmentAccepted.RecordedOn;
                acceptedEnrollment.StatusReason = enrollmentAccepted.ReasonDescription;
                _enrollmentRepository.Save(acceptedEnrollment);
                break;

            case EnrollmentDeclined enrollmentDeclined:
                var declinedEnrollment = _enrollmentRepository.FindOneById(enrollmentDeclined.AggregateId) ?? throw new InvalidOperationException();
                declinedEnrollment.Status = EnrollmentStatus.Declined.ToString();
                declinedEnrollment.ReviewedOn = enrollmentDeclined.RecordedOn;
                declinedEnrollment.StatusReason = enrollmentDeclined.ReasonDescription;
                _enrollmentRepository.Save(declinedEnrollment);
                break;
        }
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/projection/enrollmentList/EnrollmentListProjectionHandler.ts
import { ProjectionHandler } from '../../../../common/projection/ProjectionHandler';
import { Enrollment } from './Enrollment';
import { EnrollmentRepository } from './EnrollmentRepository';
import { EnrollmentRequested } from '../../event/EnrollmentRequested';
import { EnrollmentAccepted } from '../../event/EnrollmentAccepted';
import { EnrollmentDeclined } from '../../event/EnrollmentDeclined';
import {ProductDefined} from "../../../product/event/ProductDefined";
import {ProductNameRepository} from "./ProductNameRepository";
import {ProductName} from "./ProductName";
import {EnrollmentStatus} from "../../aggregate/EnrollmentStatus";

export class EnrollmentListProjectionHandler extends ProjectionHandler {
    constructor(
        private readonly enrollmentRepository: EnrollmentRepository,
        private readonly productNameRepository: ProductNameRepository
    ) {
        super();
    }

    async project(event: any): Promise<void> {
        if (event instanceof ProductDefined) {
            await this.productNameRepository.save(new ProductName(event.aggregateId, event.name));
        } else if (event instanceof EnrollmentRequested) {
            const enrollment = new Enrollment(
                event.aggregateId,
                event.userId,
                event.productId,
                event.recordedOn,
                EnrollmentStatus.Requested
            );
            await this.enrollmentRepository.save(enrollment);
        } else if (event instanceof EnrollmentAccepted) {
            const enrollment = await this.enrollmentRepository.findOneById(event.aggregateId);
            if (!enrollment) throw new Error('Enrollment not found');
            enrollment.status = EnrollmentStatus.Accepted;
            enrollment.reviewedOn = event.recordedOn;
            enrollment.statusReason = event.reasonDescription;
            await this.enrollmentRepository.save(enrollment);
        } else if (event instanceof EnrollmentDeclined) {
            const enrollment = await this.enrollmentRepository.findOneById(event.aggregateId);
            if (!enrollment) throw new Error('Enrollment not found');
            enrollment.status = EnrollmentStatus.Declined;
            enrollment.reviewedOn = event.recordedOn;
            enrollment.statusReason = event.reasonDescription;
            await this.enrollmentRepository.save(enrollment);
        }
    }
}